server/utils/APIWrapper.ts

```diff
@@ -86,4 +151,95 @@ function APIWrapper(
   };
 }
 
-export default APIWrapper;
+const ajvBodyParser = new Ajv();
+const ajvQueryParamParser = new Ajv();
+
+function validateAndCoerceReq<Q_P, B, R>(
+  { config, handler }: Route<Q_P, B, R>,
+  req: InternalRequest<StringsDict, UnknownDict>
+): InternalRequest<Q_P, B> {
+  const castedReq = req as unknown as InternalRequest<Q_P, B>;
+  castedReq.query = config?.querySchema
+    ? validateAndCoerceDataToSchema(
+        "query",
+        ajvQueryParamParser,
+        config?.querySchema,
+        parseJSONQueryParams(config.querySchema, req.query)
+      )
+    : (req.query as unknown as Q_P);
+  castedReq.body = config?.bodySchema
+    ? validateAndCoerceDataToSchema(
+        "body",
+        ajvBodyParser,
+        config.bodySchema,
+        req.body
+      )
+    : (req.body as unknown as B);
+  return castedReq;
+}
+
+// parseJSONQueryParams attempts to convert strings that are expected to be objects from JSON.
+// it infers if a param type is object from the schema. if this does not work consistently enough,
+// then switch it over from inference to a configuration option specifying the params that need to be transformed
+// or create a custom type that wraps a schema and specifies root level transformations in that wrapper
+function parseJSONQueryParams(
```

Comment:
Can you explain this a bit more?  I am not understanding what situation this is accounting for.

Reply:
Yeah this is used when validating and coercing object params in the query params since query params can only be `string | string[]`. ajv can't really coerce JSON strings into objects without a lot of hassle, so this method essentially expands the types to `string | string[] | Object` if that makes sense

E.g., `/blah?filters={"someKey":"someValue", "someOtherKey": "someOtherValue"}

--> query params dict parsed by Next.Js: `{filters: "a string of the above JSON"}`

--> query params dict parsed by this method: `{filters: {someKey: "someValue", someOtherKey: "someOtherValue"}`